### PR TITLE
Fixed size shapes

### PR DESCRIPTION
**Pull request by rmoestl**
This implements *fixed size shapes* specified in #2193. Please refer to the commit message for a rough overview of the feature.

### CODE REVIEW

File: src/components/shapes/attributes.js [hunk 1]
```diff
@@ -71,13 +75,41 @@ module.exports = {
             'the date to unix time in milliseconds.'
         ].join(' ')
     }),
+    xsizemode: {
+        valType: 'enumerated',
+        values: ['data', 'pixel'],
+        dflt: 'data',
```
Comment (alexcjohnson) [inline src/components/shapes/attributes.js]: Ah, I hadn't thought of this before, but `'data'` mode applies not just to axis scaling but also plot fraction scaling, right? ie when `xref='paper'` this option is still relevant, right? So how about we change `'data'` to `'scaled'`? That seems like it encompasses both cases and differentiates them from `'pixel'` sizing...

Comment (rmoestl) [reply src/components/shapes/attributes.js]: Good point. I'll change that.

**Comment by rmoestl** [comment]
The new test image is here: https://github.com/plotly/plotly.js/blob/fixed-size-shapes/test/image/baselines/fixed_size_shapes.png

File: test/image/mocks/fixed_size_shapes.json [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        "line": {
+          "color": "rgba(96, 171, 50, 1)"
+        }
+      }
+    ]
+  },
+  "config": {
+    "editable": "true"
+  }
+}
```
Comment (alexcjohnson) [inline test/image/mocks/fixed_size_shapes.json]: Very nice test image, and I think the interactions feel very natural as you have them. Other than adding a trailing newline to make the [syntax test happy](https://circleci.com/gh/plotly/plotly.js/8691?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) the only change I'd like to see is to use the other three axis types (log, date, category) on one axis each - no need to add more subplots, just convert a couple of the existing axes.

**Comment by alexcjohnson** [comment]
One thing I just noticed is the cursors start out wrong (at least in some cases) - the "top" cursor shows for most (or all!) of the shape, then after you cause a redraw (in the gif below I zoom in then double click to autorange again) the cursors look correct.

I haven't looked to see if this is exclusive to pixel-sized shapes, nor if it's new with this PR... can you check it out @rmoestl ?

weird shapes cursor

**Comment by rmoestl** [comment]
I'll check out the cursor position thing.

**Comment by rmoestl** [comment]
Re: drag cursors, unfortunately I was not yet able to reproduce this strange behavior. 

Looking at the code, the bounding box of the shape is obtained in `setupDragElement` and later used in the mouse move event handler to calculate the cursor. Can resizing cause this behavior? I don't think so because a resize triggers drawing the shape which in turn triggers a new call of `setupDragElement`. In addition your recording shows that you've not yet resized the respective shape.

Do you have some hints on how to reproduce the bug? I was testing in Chrome and FF by the way. Though I assume `getBoundingClientRect` is working correct in all major browsers.

**Comment by alexcjohnson** [comment]
Thanks for the test image tweaks - love it!

I figured out what the bug is - it's when you scroll the page after the plot draws (which I was doing naturally with your mock because it's fairly tall, whereas the original `shapes` mock is much shorter - if you shrink your browser window so you need to scroll that one you can see it there too).

It's not new in this PR, but I think it has a simple solution so you might as well address it here. The issue is that `getBoundingClientRect` gets called in the outer scope of `setupDragElement`, and gets used later in `updateDragMode`, at which point the actual bbox is different if you've scrolled. The simple solution is to move `getBoundingClientRect` to the scope where it's used.

**Comment by rmoestl** [comment]
Wow, good catch!!

**Comment by alexcjohnson** [comment]
Great, glad that's all it took! That bug fix may actually help performance too - ironically, since now we're calling `getBoundingClientRect` many more times, but we're *not* calling it during draw, only during mouse moves.

I'm happy with this now, lets go! 💃

**Comment by Juanlu001** [comment]
Thanks for this new feature! As a Python user, I understand that the `offline` mode always uses the latest release from the CDN so this should be ready to use. Have the online docs been update accordingly?

**Comment by rmoestl** [comment]
@Juanlu001 ICYMI, online docs have been updated: https://plot.ly/javascript/reference/#layout-shapes

**Pull request merged**